### PR TITLE
Bug fix + Update + Refact for Remove text for HI + TEST

### PR DESCRIPTION
**Pull request by ivandrofly**
**All test passed** /cc **@niksedk**

### CODE REVIEW

File: src/Logic/Forms/RemoveTextForHI.cs [hunk 1]
```diff
@@ -852,45 +865,38 @@ private string RemoveStartEndTags(string text)
...
         {
             string preAssTag = string.Empty;
-            if (text.StartsWith("{\\") && text.IndexOf('}') > 0)
+            if (text.StartsWith("{\\", StringComparison.Ordinal) && text.IndexOf('}') > 0)
             {
                 int indexOfEndBracket = text.IndexOf('}') + 1;
                 preAssTag = text.Substring(0, indexOfEndBracket);
                 text = text.Remove(0, indexOfEndBracket).TrimStart();
             }
-
```
Comment (ivandrofly) [inline src/Logic/Forms/RemoveTextForHI.cs]: I don't this this is needed...

File: src/Logic/Forms/RemoveTextForHI.cs [hunk 2]
```diff
@@ -80,78 +80,80 @@ public string RemoveColon(string text)
                 text = text.Remove(0, indexOfEndBracket).TrimStart();
             }
 
-            if (!text.Contains(':'))
```
Comment (ivandrofly) [inline src/Logic/Forms/RemoveTextForHI.cs]: Why not checking *(in line # 72)* before stripping **AssTags**

File: src/Logic/Forms/RemoveTextForHI.cs [hunk 3]
```diff
@@ -80,78 +80,80 @@ public string RemoveColon(string text)
...
             // House 7x01 line 52: and she would like you to do three things:
             // Okay or remove???
-            if (text.IndexOf(':') > 0 && text.IndexOf(':') == text.Length - 1 && text != text.ToUpper())
+            var noTagText = HtmlUtil.RemoveHtmlTags(text);
+            if (noTagText.IndexOf(':') > 0 && noTagText.IndexOf(':') == noTagText.Length - 1 && noTagText != noTagText.ToUpper())
                 return text;
 
             string newText = string.Empty;
-            var parts = text.Trim().SplitToLines();
+            var lines = text.Trim().SplitToLines();
```
Comment (ivandrofly) [inline src/Logic/Forms/RemoveTextForHI.cs]: I think it's more readable this way :sunglasses:

File: src/Logic/Forms/RemoveTextForHI.cs [hunk 4]
```diff
@@ -80,78 +80,80 @@ public string RemoveColon(string text)
...
-                                newText = newText + Environment.NewLine + "<b>" + s;
-                            else if (pre.Contains('[') && s.Contains(']'))
-                                newText = newText + Environment.NewLine + "[" + s;
-                            else if (pre.Contains('(') && s.EndsWith(')'))
-                                newText = newText + Environment.NewLine + "(" + s;
+                            if (pre.Contains("<i>") && line.Contains("</i>"))
+                                newText = newText + Environment.NewLine + "<i>" + line;
+                            else if (pre.Contains("<b>") && line.Contains("</b>"))
+                                newText = newText + Environment.NewLine + "<b>" + line;
+                            else if (pre.Contains("<u>") && line.Contains("</u>"))
```
Comment (ivandrofly) [inline src/Logic/Forms/RemoveTextForHI.cs]: Include <code>\<u></code> tag

File: src/Logic/Forms/RemoveTextForHI.cs [hunk 5]
```diff
@@ -80,78 +80,80 @@ public string RemoveColon(string text)
...
-                            if (indexOfColon > 0 && indexOfColon < s.Length - 1)
+                            if (indexOfColon > 0 && indexOfColon < line.Length - 1)
                             {
-                                if (char.IsDigit(s[indexOfColon - 1]) && char.IsDigit(s[indexOfColon + 1]))
-                                    remove = false;
+                                remove = !Utilities.IsBetweenNumbers(line, indexOfColon);
                             }
-                            if (s.StartsWith("Previously on") || s.StartsWith("<i>Previously on"))
+
+                            if (!DoRemove(pre))
```
Comment (ivandrofly) [inline src/Logic/Forms/RemoveTextForHI.cs]: Check if contains any words which shouldn't be remove here...

File: src/Logic/Forms/RemoveTextForHI.cs [hunk 6]
```diff
@@ -182,27 +184,32 @@ public string RemoveColon(string text)
 
                                 if (text.StartsWith('(') && newText.EndsWith(')') && !newText.Contains('('))
                                     newText = newText.TrimEnd(')');
-                                else if (newText.EndsWith("</i>") && text.StartsWith("<i>") && !newText.StartsWith("<i>"))
+                                else if (text.StartsWith('[') && newText.EndsWith(']') && !newText.Contains('['))
+                                    newText = newText.TrimEnd(']');
+                                else if (newText.EndsWith("</i>", StringComparison.Ordinal) && text.StartsWith("<i>", StringComparison.Ordinal) && !newText.StartsWith("<i>", StringComparison.Ordinal))
                                     newText = "<i>" + newText;
-                                else if (newText.EndsWith("</b>") && text.StartsWith("<b>") && !newText.StartsWith("<b>"))
+                                else if (newText.EndsWith("</b>", StringComparison.Ordinal) && text.StartsWith("<b>", StringComparison.Ordinal) && !newText.StartsWith("<b>", StringComparison.Ordinal))
```
Comment (ivandrofly) [inline src/Logic/Forms/RemoveTextForHI.cs]: Added <code>StringComparison.Ordinal</code>

File: src/Logic/Forms/RemoveTextForHI.cs [hunk 7]
```diff
@@ -212,14 +219,7 @@ public string RemoveColon(string text)
...
                                     {
-                                        int periodIndex = start.LastIndexOf(". ", StringComparison.Ordinal);
-                                        int questIndex = start.LastIndexOf("? ", StringComparison.Ordinal);
-                                        int exclaIndex = start.LastIndexOf("! ", StringComparison.Ordinal);
-                                        int endIndex = periodIndex;
-                                        if (questIndex > endIndex)
-                                            endIndex = questIndex;
-                                        if (exclaIndex > endIndex)
-                                            endIndex = exclaIndex;
+                                        int endIndex = start.LastIndexOfAny(new[] { '.', '!', '?' });
```
Comment (ivandrofly) [inline src/Logic/Forms/RemoveTextForHI.cs]: I think this is better

File: src/Logic/Forms/RemoveTextForHI.cs [hunk 8]
```diff
@@ -290,7 +289,7 @@ public string RemoveColon(string text)
                 {
                     if (indexOfDialogChar < 0 || indexOfDialogChar > 4)
                     {
-                        var st = new StripableText(newText, "", "");
+                        var st = new StripableText(newText, string.Empty, string.Empty);
```
Comment (ivandrofly) [inline src/Logic/Forms/RemoveTextForHI.cs]: <code>String.Empty</code> shouldn't be confused as "" which may contain some weird unicode whitespace

File: src/Logic/Forms/RemoveTextForHI.cs [hunk 9]
```diff
@@ -527,12 +538,13 @@ private bool IsHIDescription(string text)
             if (text.Trim().Replace("mr. ", string.Empty).Replace("mrs. ", string.Empty).Replace("dr. ", string.Empty).Contains(' '))
                 AddWarning();
 
-            if (text == "sighing" ||
```
Comment (ivandrofly) [inline src/Logic/Forms/RemoveTextForHI.cs]: Will be handled in line **573**

File: src/Logic/Forms/RemoveTextForHI.cs [hunk 10]
```diff
@@ -962,16 +966,16 @@ public string RemoveLineIfAllUppercase(string text)
             if (!Settings.RemoveIfAllUppercase)
                 return text;
 
-            string[] lines = text.Replace(Environment.NewLine, "\n").Replace("\r", "\n").Split('\n');
+            var lines = text.SplitToLines();
```
Comment (ivandrofly) [inline src/Logic/Forms/RemoveTextForHI.cs]: Use SE's new <code>SplitToLines();</code>

File: src/Logic/Forms/RemoveTextForHI.cs [hunk 11]
```diff
@@ -962,16 +966,16 @@ public string RemoveLineIfAllUppercase(string text)
...
 
-            string[] lines = text.Replace(Environment.NewLine, "\n").Replace("\r", "\n").Split('\n');
+            var lines = text.SplitToLines();
             var sb = new StringBuilder();
-            foreach (string line in lines)
+            foreach (var line in lines)
             {
-                string lineNoHtml = HtmlUtil.RemoveHtmlTags(line);
-                string tmp = lineNoHtml.TrimEnd('.', '!', '?', ':').Trim();
-                if (lineNoHtml != lineNoHtml.ToLower() && lineNoHtml == lineNoHtml.ToUpper())
```
Comment (ivandrofly) [inline src/Logic/Forms/RemoveTextForHI.cs]: I think check if <code>lineNoHtml == lineNoHtml.ToUpper()</code> is enough

**Comment by niksedk** [comment]
:)

**Pull request merged**